File: Leetcode/2149. Rearrange Array Elements by Sign.py
# link - https://leetcode.com/problems/rearrange-array-elements-by-sign/

# space: O(n)
# time: O(n)

class Solution:
    def rearrangeArray(self, nums: List[int]) -> List[int]:
        pos, neg, ans = 0, 1, [None] * len(nums)
        for num in nums:
            if num > 0:
                ans[pos] = num
                pos += 2
            else:
                ans[neg] = num
                neg += 2
        return ans

File: Leetcode/2149. Rearrange Array Elements by Sign.cpp
/*
    https://leetcode.com/problems/rearrange-array-elements-by-sign/
    Time: O(n)
    Space: O(n)
*/

class Solution {
public:
    vector<int> rearrangeArray(vector<int> &nums) {

        int pos = 0, neg = 0, turn = 1;
        vector<int> res;

        while (res.size() < nums.size()) {
            if (turn == 1) {
                while (nums[pos] < 0)
                    pos++;
                res.push_back(nums[pos]);
                pos++;
            } else {
                while (nums[neg] > 0)
                    neg++;
                res.push_back(nums[neg]);
                neg++;
            }
            turn = (turn + 1) % 2;
        }

        return res;
    }
};